staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -0,0 +1,261 @@
+
+	server := newApfServer(decisionReject, t)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusTooManyRequests); err != nil {
+		t.Error(err)
+	}
+
+	checkForExpectedMetricsWithRetry(t, []string{
+		"apiserver_request_terminations_total",
+		"apiserver_dropped_requests_total",
+	})
+}
+
+func TestApfExemptRequest(t *testing.T) {
+	epmetrics.Register()
+	fcmetrics.Register()
+
+	// wait the first sampleAndWaterMark metrics to be collected
+	time.Sleep(time.Millisecond * 50)
+
+	server := newApfServer(decisionNoQueuingExecute, t)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusOK); err != nil {
+		t.Error(err)
+	}
+
+	checkForExpectedMetricsWithRetry(t, []string{
+		"apiserver_current_inflight_requests",
+		"apiserver_flowcontrol_read_vs_write_request_count_watermarks",
+		"apiserver_flowcontrol_read_vs_write_request_count_samples",
+	})
+}
+
+func TestApfExecuteRequest(t *testing.T) {
+	epmetrics.Register()
+	fcmetrics.Register()
+
+	// wait the first sampleAndWaterMark metrics to be collected
+	time.Sleep(time.Millisecond * 50)
+
+	server := newApfServer(decisionQueuingExecute, t)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusOK); err != nil {
+		t.Error(err)
+	}
+
+	checkForExpectedMetricsWithRetry(t, []string{
+		"apiserver_current_inflight_requests",
+		"apiserver_current_inqueue_requests",
+		"apiserver_flowcontrol_read_vs_write_request_count_watermarks",
+		"apiserver_flowcontrol_read_vs_write_request_count_samples",
+	})
+}
+
+func TestApfCancelWaitRequest(t *testing.T) {
+	epmetrics.Register()
+
+	server := newApfServer(decisionCancelWait, t)
+	defer server.Close()
+
+	if err := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusTooManyRequests); err != nil {
+		t.Error(err)
+	}
+
+	checkForExpectedMetricsWithRetry(t, []string{
+		"apiserver_current_inflight_requests",
+		"apiserver_request_terminations_total",
+		"apiserver_dropped_requests_total",
+	})
+}
+
+// wait async metrics to be collected
+func checkForExpectedMetricsWithRetry(t *testing.T, expectedMetrics []string) {
+	maxRetries := 5
+	var checkErrors []error
+	for i := 0; i < maxRetries; i++ {
+		t.Logf("Check for expected metrics with retry %d", i)
+		metricsFamily, err := legacyregistry.DefaultGatherer.Gather()
+		if err != nil {
+			t.Fatalf("Failed to gather metrics %v", err)
+		}
+
+		metrics := map[string]interface{}{}
+		for _, mf := range metricsFamily {
+			mf := mf
+			metrics[*mf.Name] = mf
+		}
+
+		checkErrors = checkForExpectedMetrics(expectedMetrics, metrics)
+		if checkErrors == nil {
+			return
+		}
+
+		time.Sleep(1 * time.Second)
+	}
+	for _, checkError := range checkErrors {
+		t.Error(checkError)
+	}
+}
+
+func checkForExpectedMetrics(expectedMetrics []string, metrics map[string]interface{}) []error {
```

Comment:
is the expectation that the metrics will not exist if they didn't get set at least once? can check for explicit values? we can reset the metrics at the beginning of each test I think (or assert on some delta).

Reply:
Yes, Prometheus `HistogramVector` objects are being used, which means the actual metrics do not exist until data is put into at least one Histogram in the Vector.

Reply:
can we check for specific values instead of just checking that the metric exist?

Reply:
yes, if the test func waits for at least one sampling period, after any activity, before reading the metrics.